### PR TITLE
add_cloud_metadata - Retrieve AKS cluster name and id

### PR DESCRIPTION
**Pull request by MichaelKatsoulis**
## Proposed commit message

- WHAT: Enhance `add_cloud_metadata` processor with `orchestrator.cluster.name` and `orchestrator.cluster.id` for AKS clusters.
- WHY:  It allows filtering of events based on the AKS cluster the vm belongs to.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

- Create an AKS cluster
- Run Metricbeat or Filebeat with [sleep](https://github.com/elastic/beats/blob/90228e66040f055f8bcd5395e7161a3d3f26018e/metricbeat/module/kubernetes/_meta/test/docs/01_playground/metricbeat.yaml#L155) command and correct values of following additional env vars
 ```
- name: TENANT_ID
  value: ""
- name: CLIENT_ID
  value: ""
- name: CLIENT_SECRET
  value: ""
 ```
In order to acquire a CLIENT_SECRET and CLIENT_ID follow the [app_registration](https://docs.elastic.co/integrations/azure_billing#app-registration) instructions.
- Build this pr's version of metricbeat/filebeat with `GOOS=linux GOARCH=amd64 go build`
- Copy the metricbeat/filebeat binary in metricbeat/filebeat pod with `kubectl cp ./metricbeat metricbeat-9lkhf:/usr/share/metricbeat/ -n kube-system`
- Exec in the pod and run `./metricbeat  -c /etc/metricbeat.yml -e`
- Check in Kibana the fields `orchestrator.cluster.name` and `orchestrator.cluster.id` being populated

## Related issues

- Closes https://github.com/elastic/beats/issues/33081

## Screenshots

orchestrator

### CODE REVIEW

File: libbeat/processors/add_cloud_metadata/provider_azure_vm.go
Comment (gizas) [inline libbeat/processors/add_cloud_metadata/provider_azure_vm.go:120]: Can you add a comment for the fetchMetadata fucntion before the definiton?

File: libbeat/processors/add_cloud_metadata/provider_azure_vm.go
Comment (gizas) [inline libbeat/processors/add_cloud_metadata/provider_azure_vm.go:107]: Defined in line 44. Can you also add a comment in line 44 or here?
Basic some explanation why hfetcher, fetcher and gfetcher differ?
sth like :
hfetcher: creates the handler
fetcher: the handler with the specific cloud provider is initialised
gfethcer: the actual metadata fetcher that calls fetchAzureClusterMeta

**Comment by gizas** [comment]
My E2e tests are su
Screenshot 2024-02-05 at 5 30 43 PM
ccessful and also the tests you had included
Screenshot 2024-02-05 at 5 28 59 PM


We will need to update:
- https://www.elastic.co/guide/en/fleet/8.12/running-on-aks-managed-by-fleet.html and remove the comment for add_field. Also information that TENANT_ID, CLIENT_ID and CLIENT_SECRET need to be present you must define them. From the code I also see SUBSCRIPTION_ID as optional 

- Additional pages to be updated: https://github.com/elastic/beats/blob/d778f993f5a75ec2fd08fa016465a15ab25cf04d/libbeat/processors/add_cloud_metadata/docs/add_cloud_metadata.asciidoc with Azure example

File: libbeat/processors/add_cloud_metadata/provider_azure_vm.go
Comment (gizas) [inline libbeat/processors/add_cloud_metadata/provider_azure_vm.go:190]:
```suggestion
				if strSubscriptionId = os.Getenv("SUBSCRIPTION_ID"); strSubscriptionId == "" {
```

**Review by gizas — APPROVED** [review]

File: libbeat/processors/add_cloud_metadata/provider_azure_vm.go
Comment (tetianakravchenko) [inline libbeat/processors/add_cloud_metadata/provider_azure_vm.go:187]: in what cases `subscriptionId` might not be retrieved from metadata? 
we are also not mentioning anywhere that `SUBSCRIPTION_ID` can/should be provided as env var

**Review by tetianakravchenko — COMMENTED** [review]
please add a changelog regarding the `orchestrator.*` and `resourceGroup` fields

File: libbeat/processors/add_cloud_metadata/docs/add_cloud_metadata.asciidoc
Comment (zmoog) [inline libbeat/processors/add_cloud_metadata/docs/add_cloud_metadata.asciidoc:56]: Using `azidentity.NewDefaultAzureCredential(nil)` as a fallback when there ir no environment variables is a good call. 

`azidentity.NewDefaultAzureCredential(nil)` gives more options to the users that want and have the knowledge to use different authentication mechanisms like the Workload Identity [we experimented](https://github.com/zmoog/public-notes/issues/72) with.

What about mentioning this opportunity in the docs?

**Review by zmoog — APPROVED** [review]
LGTM, I would consider mentioning workload identity in the docs.

File: libbeat/processors/add_cloud_metadata/provider_azure_vm.go
Comment (MichaelKatsoulis) [reply libbeat/processors/add_cloud_metadata/provider_azure_vm.go:187]: As far as I have seen, in no way. I just wanted to give an extra option. The user may check the logs and find that they have an alternative. I am not sure where to document this without getting in so many details of the implementation

Comment (tetianakravchenko) [reply libbeat/processors/add_cloud_metadata/provider_azure_vm.go:187]: > I am not sure where to document this without getting in so many details of the implementation

I see
I still think that an extra option might be unnecessary and add more confusing - providing hidden alternative, that solves unknown issue

but it is not a blocker for this PR

Comment (MichaelKatsoulis) [reply libbeat/processors/add_cloud_metadata/provider_azure_vm.go:187]: I will remove the env var option as it may be confusing.

**Comment by MichaelKatsoulis** [comment]
@tetianakravchenko , @gizas , @zmoog  I addressed all comments. Could you take a final look?
Also I had to declare in processors fields the new `cloud.resourceGroup.name` field as it is not an ECS field like most of cloud fields. The same has beed done in the past for `cloud.image.id` as an example.

File: CHANGELOG.next.asciidoc
Comment (tetianakravchenko) [inline CHANGELOG.next.asciidoc:134]:
```suggestion
- Enhance add_cloud_metadata processor with `orchestrator.cluster.name`, `orchestrator.cluster.id` and `cloud.resource_group.name` when running inside an AKS cluster. {issue}33081[33081] {pull}37685[37685]
```

File: libbeat/processors/add_cloud_metadata/docs/add_cloud_metadata.asciidoc
Comment (tetianakravchenko) [inline libbeat/processors/add_cloud_metadata/docs/add_cloud_metadata.asciidoc:168]: name of the field should be `resource_group`?

Comment (MichaelKatsoulis) [reply libbeat/processors/add_cloud_metadata/docs/add_cloud_metadata.asciidoc:168]: Yes. Fixing this

**Comment by MichaelKatsoulis** [comment]
@elastic/elastic-agent  could I get a review on this as you are the code owners ?

**Comment by axw** [comment]
> The reason I named the field cloud.resource_group.name is to follow the Otel [schema](https://opentelemetry.io/docs/specs/semconv/attributes-registry/cloud/) (you can see the cloud.resource_id field). I made the cloud.resource_group to be a group variable with name as a field because it could possibly include other fields in the future such as type.

Cloud provider-specific attributes are typically prefixed with the name of the cloud provider. For Azure, there's some prior art here: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/resourcedetectionprocessor#azure

**Comment by MichaelKatsoulis** [comment]
>Cloud provider-specific attributes are typically prefixed with the name of the cloud provider. For Azure, there's some prior art here: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/resourcedetectionprocessor#azure

@axw I wouldn't consider resource_group.name as an azure cloud provider specific field , although it is only populated by azure cloud metadata processor for now. It is relevant for all providers. ([GCP](https://cloud.google.com/monitoring/groups), [AWS](https://docs.aws.amazon.com/ARG/latest/userguide/resource-groups.html))

**Comment by axw** [comment]
@MichaelKatsoulis that's fair, but my point was more: if we want to align with OTel semantic conventions, then there isn't any one for resource group; there is however an Azure-specific one.

**Comment by MichaelKatsoulis** [comment]
> there is however an Azure-specific one.

@axw I couldn't find it in https://opentelemetry.io/docs/specs/semconv/

**Review by tetianakravchenko — APPROVED** [review]

**Comment by MichaelKatsoulis** [comment]
@axw ohh I see! Thanks for pointing this out. I update the field name to be aligned with resource detection processor!

File: libbeat/processors/add_cloud_metadata/provider_azure_vm.go
Comment (fearful-symmetry) [inline libbeat/processors/add_cloud_metadata/provider_azure_vm.go:109]: Can we add `fmt.Errorf` statements to the errors here? We put them in some of the later functions, so we should be consistent.

**Review by aleksmaus — APPROVED** [review]

File: libbeat/processors/add_cloud_metadata/provider_azure_vm.go
Comment (MichaelKatsoulis) [reply libbeat/processors/add_cloud_metadata/provider_azure_vm.go]: @fearful-symmetry Done

**Review by fearful-symmetry — APPROVED** [review]

File: libbeat/processors/add_cloud_metadata/_meta/fields.yml
Comment (andrewkroh) [inline libbeat/processors/add_cloud_metadata/_meta/fields.yml:55]: Is this consistent with what we have been using in Fleet integrations? I think I recall `azure.resource.group` being used over there.

File: libbeat/processors/add_cloud_metadata/docs/add_cloud_metadata.asciidoc
Comment (andrewkroh) [inline libbeat/processors/add_cloud_metadata/docs/add_cloud_metadata.asciidoc:58]: Why is "Virtual Machine" treated as a proper name?

```suggestion
   If the virtual machine is part of an AKS managed cluster, the fields
```

File: libbeat/processors/add_cloud_metadata/docs/add_cloud_metadata.asciidoc
Comment (andrewkroh) [inline libbeat/processors/add_cloud_metadata/docs/add_cloud_metadata.asciidoc:62]:
```suggestion
   fallback to https://learn.microsoft.com/en-us/azure/developer/go/azure-sdk-authentication?tabs=bash#2-authenticate-with-azure[DefaultAzureCredential] and user can choose different authentication methods (e.g. workload identity).
```

File: libbeat/processors/add_cloud_metadata/docs/add_cloud_metadata.asciidoc
Comment (andrewkroh) [inline libbeat/processors/add_cloud_metadata/docs/add_cloud_metadata.asciidoc:163]: Field names should not contain dots when emitted from Beats. Restructure the event so that it produces `{azure: {resourcegroup: {name: ...`

update: I didn't see any incorrect code. So I think this is just a doc issue.

File: libbeat/processors/add_cloud_metadata/provider_azure_vm.go
Comment (andrewkroh) [inline libbeat/processors/add_cloud_metadata/provider_azure_vm.go:57]: The type should be omitted.

```suggestion
var NewClusterClient = func(clientFactory *armcontainerservice.ClientFactory) *armcontainerservice.ManagedClustersClient {
```

File: libbeat/processors/add_cloud_metadata/provider_azure_vm.go
Comment (andrewkroh) [inline libbeat/processors/add_cloud_metadata/provider_azure_vm.go:56]: The docs don't match the code (`NewManagedClustersClient` is not returned).

File: libbeat/processors/add_cloud_metadata/provider_azure_vm.go
Comment (andrewkroh) [inline libbeat/processors/add_cloud_metadata/provider_azure_vm.go:133]: To make this render nicely in godoc, indent these two items.

```suggestion
//   1. Azure metadata endpoint with httpMetadataFetcher
//   2. Azure Managed Clusters using azure sdk  with genericMetadataFetcher
```

File: libbeat/processors/add_cloud_metadata/provider_azure_vm.go
Comment (andrewkroh) [inline libbeat/processors/add_cloud_metadata/provider_azure_vm.go:167]: Remove `logger` as a param if its not used.

File: libbeat/processors/add_cloud_metadata/provider_azure_vm.go
Comment (andrewkroh) [inline libbeat/processors/add_cloud_metadata/provider_azure_vm.go:111]:
```suggestion
		// fetcher represents an azure metadata fetcher. The struct includes two types of fetchers.
```

File: libbeat/processors/add_cloud_metadata/provider_azure_vm.go
Comment (andrewkroh) [inline libbeat/processors/add_cloud_metadata/provider_azure_vm.go:166]:
```suggestion
// getAKSClusterNameId returns the AKS cluster name and ID for a given resourceGroup
```

Can you please do the same change everywhere (like in error messages). Either `ID`, `id`, or `identifier` are good.

Comment (andrewkroh) [reply libbeat/processors/add_cloud_metadata/provider_azure_vm.go:166]: In the code as well, I recommend to change `Id` to `ID`. It is the Go convention. See https://go.dev/wiki/CodeReviewComments#initialisms.

File: libbeat/processors/add_cloud_metadata/_meta/fields.yml
Comment (andrewkroh) [reply libbeat/processors/add_cloud_metadata/_meta/fields.yml:55]: I am also a little concerned about this processor beginning to produce fields outside of the ECS `cloud` and `orchestrator` namespaces.

Downstream consumers (e.g Fleet integrations) should ideally be updated to have this field included in their mappings since it's not part of ECS. This affects almost all integrations since Agent includes `add_cloud_metadata` with most input types from Beats. (Imagine if we ran our CI on Azure for Fleet integration system testing, after this merges every integration's system tests would begin failing because this field was not defined in its mappings.)

The other concern is whether or not existing pipelines will safely handle the addition of this field. For example if an ingest pipeline is already using this field then does it safely handle the case where the field already exists.

Comment (MichaelKatsoulis) [reply libbeat/processors/add_cloud_metadata/_meta/fields.yml:55]: The `azure.resource.group` , created by a pipeline? as I noticed is the resource group stripped. It only includes the group's name. The addition of the `azure.resourcegroup.name` field I did was to me a nice to have and definitely not the reason of this PR.
I understand that it then needs to be added in all the integrations because it comes from the processor.
Also it is confusing having two similar named fields. I chose the naming to be compliant with Otel.

I hear your concerns and I will remove this field as it beats the purpose of this PR which was to add the orchestrator ecs fields.

File: libbeat/processors/add_cloud_metadata/provider_azure_vm.go
Comment (andrewkroh) [inline libbeat/processors/add_cloud_metadata/provider_azure_vm.go:56]:
```suggestion
// NewClusterClient variable is assigned an anonymous function that returns a new ManagedClustersClient.
```

File: libbeat/processors/add_cloud_metadata/_meta/fields.yml
Comment (andrewkroh) [reply libbeat/processors/add_cloud_metadata/_meta/fields.yml]: > The azure.resource.group , created by a pipeline?

I'm not familiar with these integrations, but, yes, I see it being created by a grok in a few pipeline.

https://github.com/search?q=repo%3Aelastic%2Fintegrations+path%3A**%2F*.yml+%2Fazure%5C.resource%5C.%3Fgroup%2F&type=code

> I understand that it then needs to be added in all the integrations because it comes from the processor.

This is a really general problem we have with Fleet integrations in that there is no easy or centralized way to account for the fields being added by the processors.

> I hear your concerns and I will remove this field as it beats the purpose of this PR which was to add the orchestrator ecs fields.

👍

**Review by andrewkroh — APPROVED** [review]

**Review by pierrehilbert — APPROVED** [review]

**Comment by fxdgear** [comment]
> In order to acquire a CLIENT_SECRET and CLIENT_ID follow the [app_registration](https://docs.elastic.co/integrations/azure_billing#app-registration) instructions.

What role does the app need in order to get this information?

**Comment by maggieghamry** [comment]
@MichaelKatsoulis is it possible to add the version label this was merged into, and if there were any backported versions? The release notes seem to indicate this was for `8.13.0+` , is that correct?

**Comment by MichaelKatsoulis** [comment]
@maggieghamry  It was added in 8.14.0 and there were no backports.   I don't see any label available for 8.14.0 to add.

I just checked the release notes and they say that it was added in 8.13.0. I don't understand how this has happened. They are clearly wrong. Maybe they were broken due to some backported PR.

This feature is part of `8.14.0+`.

**Comment by maggieghamry** [comment]
Thank you @MichaelKatsoulis , yes it looks like because of https://www.elastic.co/guide/en/beats/libbeat/current/release-notes-8.13.0.html#_added_6 it is creating some confusion with 8.13.x instances still experiencing the error. I think* a new label can be added with the affecting version so that it can be merged into the corresponding branch, at least so we can clearly identify which version it's added to (the PR commit doesn't specify it explicitly either), not sure if possibly the docs team can update the PR notes as well so we can try to replicate the behavior.

**Comment by ldecrescenzo** [comment]
Hi @MichaelKatsoulis! I have a customer running into this issue on 8.15.1, could you confirm if this PR was merged before this version? Thanks!

**Comment by MichaelKatsoulis** [comment]
> Hi @MichaelKatsoulis! I can see you added the 8.14 label, but I have a customer running into this issue on 8.15.1, could you confirm if this PR was merged before this version? Thanks!

I can confirm it. The code is there.

**Pull request merged**
